### ISSUE TITLE
notebbok

### ISSUE THREAD

**Issue by fourier-fans** [fourier-fans/notebook#1]
2022.3.22
1. 今天学习了paper:_Learning a Sketch Tensor Space for Image Inpainting of Man-made Scenes_,这篇paper提出了对于图像修复的任务，其auxiliary information至关重要，文中使用的canny 模块去提取图像的edge信息以及使用LSM（line segment masking）模块去提取line信息，对于LSM模块，其可能是一个net，但是具体的内容还需要查看另外一个paper： _Holisticallyattracted wireframe parsing_，之后的encoder与decoder部分使用了partially gate convulotion去优化vanilla convolution，其具体的细节也要查看相关的paper。之后的loss以及注意力机制仍然需进一步学习。
2. 之后学习了attention机制中的通道机制的训练过程，其简单学习之后可以理解，其核心在于给cnn网络中不同的通道赋予权值，其训练参数即为W矩阵。
<img width="411" alt="7CDAF8C6-7AF4-4FA8-950E-AD90EA62F36D" src="https://user-images.githubusercontent.com/91041719/159458403-ab12f9b9-deb0-48f7-9c1c-d0a1324f8317.png">
3.

**Comment by fourier-fans** [comment]
**时隔多天 我又开始写病例本了 之前傻逼疫情把我整emo了 没有好好学习 我的**

1. **gate conv** 其核心代码的思路还是非常巧妙的，其输出的是一个output_channel*2 的结果。一个output channel的输出用作gate conv的x, 一个作为g，其输出结果为结果w*g


class GateConv(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=3, stride=1, padding=1, transpose=False):
        super(GateConv, self).__init__()
        self.out_channels = out_channels
        if transpose:
            self.gate_conv = nn.ConvTranspose2d(in_channels, out_channels * 2,
                                                kernel_size=kernel_size,
                                                stride=stride, padding=padding)
        else:
            self.gate_conv = nn.Conv2d(in_channels, out_channels * 2,
                                       kernel_size=kernel_size,
                                       stride=stride, padding=padding)

    def forward(self, x):
        x = self.gate_conv(x)
        (x, g) = torch.split(x, self.out_channels, dim=1)
        return x * torch.sigmoid(g)




2.对于普通的conv module ，没什么好说的，init还是那几大要素，input_channel ,output_channel ,kernnel_size,padding等等。这里还设计一个spectral_norm的概念，这个概念可是来自于nn.utils.spectral_normd的一个函数，其输入应当是一个module。

设计到gan的概念，明天继续研究 我透
今天看到他娘的muti-attention部分 真几把搞不动了 下班

**Comment by fourier-fans** [comment]
attention模块
    def __init__(self, in_channels, dim, head_count, out_channels):
        super().__init__()
        self.in_channels = in_channels
        self.head_count = head_count
        self.dim = dim

        self.keys = nn.Conv2d(in_channels, dim, 1)
        self.queries = nn.Conv2d(in_channels, dim, 1)
        self.values = nn.Conv2d(in_channels, dim, 1)
        if dim != out_channels:
            self.reprojection = nn.Conv2d(dim, out_channels, 1)
        else:
            self.reprojection = None
这个attention模块的conv模块使用的是1*1的kernel size 这样的conv并不会改变channel 的size，为了减少computation，这里引入一个dim，但是为了保证最后的output_channel，这个reprojection进行channel恢复。


    def forward(self, input_, mask=None, return_scores=False):
        n, _, h, w = input_.size()
        keys = self.keys(input_)
        queries = self.queries(input_)
        values = self.values(input_)
        head_channels = self.dim // self.head_count

        if mask is not None:
            # [b,1,h,w]
            mask = F.interpolate(mask, size=[h, w], mode='nearest')
            keys += (mask * -10000.0)
            queries += (mask * -10000.0)

        keys = keys.reshape((n, self.dim, h * w))  # [b,d,h*w]
        queries = queries.reshape(n, self.dim, h * w)
        values = values.reshape((n, self.dim, h * w))

        attended_values = []
        scores = 0
        **for i in range(self.head_count):
            key = F.softmax(keys[:, i * head_channels: (i + 1) * head_channels, :], dim=2)
            query = F.softmax(queries[:, i * head_channels: (i + 1) * head_channels, :], dim=1)
            value = values[:, i * head_channels: (i + 1) * head_channels, :]
            context = key @ value.transpose(1, 2)  # [b, d, d]
            attended_value = (context.transpose(1, 2) @ query).reshape(n, head_channels, h, w)
            attended_values.append(attended_value)**
            if return_scores:
                score = torch.matmul(query.transpose(1, 2), key)  # [b, hw, hw]
                score = torch.mean(score, dim=1).reshape([n, h, w])
                scores += score

        aggregated_values = torch.cat(attended_values, dim=1)
        if self.reprojection is not None:
            reprojected_value = self.reprojection(aggregated_values)
        else:
            reprojected_value = aggregated_values

        attention = reprojected_value + input_

        if return_scores:
            max_value, _ = torch.max(scores.reshape([n, h * w]), dim=1)
            max_value = max_value[:, None, None]
            scores = scores / (max_value + 1e-5)
            scores = scores.unsqueeze(1)
            scores = scores.detach()
            return attention, scores
        else:
            return attention
这个forward传播，这个head_channel对应的是muti-attention的操作，这里为了减少计算量，并不适用q*k计算attention score，而是使用k*v减低维度的计算，我甚至觉得这个q直接滚了得了。。。。然后muti-attention的 直接是取出其中一部分进行计算，很怪。。。

**Comment by fourier-fans** [comment]
为了防止成果被剽窃 先拉一个坑。最近准备使用MST-inpainting的网络架构实现壁画的修复，准备从attention模块下手，借鉴华为ZITS的attention scores减少MST的computation，但是其需要在ZITS的attention score后面加一个kernel size为1 的convolution进行一个进一步的特征提取，这个功能类似于gate convolution，我称之为gate attention scores，笑死

**Comment by fourier-fans** [comment]
point 1
对于class和def的参数传递，若传输内容很多可以使用**kwagrs进行参数的传递。这样函数或类将接收到的agrs存储在kwagrs中（以字典的形式），若想要得到kwagrs中的参数，可以用for循环遍历kwagrs中的所有items（），然后用setattr(self,k,v)进行数据的存储
